src/service/action-impl.js

```diff
@@ -412,85 +445,97 @@ export class ActionService {
-        }
-
-        // Element target via `id` attribute.
-        const target = this.root_.getElementById(actionInfo.target);
+      const invokeAction = () => {
+        // Use `this.root_` as the target for global targets e.g. "AMP".
+        // Otherwise, targetType should be an element id.
+        const target = (this.globalTargets_[targetType]) ?
+          this.root_ : this.root_.getElementById(targetType);
         if (target) {
           const invocation = new ActionInvocation(target, actionInfo.method,
-              args, source, action.node, event, trust);
-          return this.invoke_(invocation, actionInfo);
+              args, source, action.node, event, trust, targetType, i);
+          return this.invoke_(invocation, action.actionInfos);
         } else {
-          this.actionInfoError_('target not found', actionInfo, target);
+          this.error_(`Target "${targetType}" not found for action ` +
+              `[${actionInfo.str}].`);
         }
       };
-
-      // Wait for the previous action, if applicable.
+      // Wait for the previous action, if any.
       currentPromise = (currentPromise)
-        ? currentPromise.then(invoke)
-        : invoke();
+        ? currentPromise.then(invokeAction)
+        : invokeAction();
     });
   }
 
   /**
-   * The errors that are a result of action definition.
-   * @param {string} s
-   * @param {?ActionInfoDef} actionInfo
-   * @param {?Element} target
+   * @param {string} message
+   * @param {?Element=} opt_element
    * @private
    */
-  actionInfoError_(s, actionInfo, target) {
-    // Method not found "activate" on ' + target
-    user().assert(false, 'Action Error: ' + s +
-        (actionInfo ? ' in [' + actionInfo.str + ']' : '') +
-        (target ? ' on [' + target + ']' : ''));
+  error_(message, opt_element) {
+    if (opt_element) {
+      // reportError() supports displaying the element in dev console.
+      const e = user().createError(`[${TAG_}] ${message}`);
+      reportError(e, opt_element);
+      throw e;
+    } else {
+      user().error(TAG_, message);
+    }
   }
 
   /**
    * @param {!ActionInvocation} invocation
-   * @param {?ActionInfoDef} actionInfo TODO(choumx): Remove this param.
+   * @param {!Array<ActionInfoDef>=} opt_actionInfos Array of infos for all
+   *   actions being invoked during a sequence (multiple actions triggered).
+   *   TODO(choumx): Remove this param by moving setState limit into this file.
    * @return {?Promise}
-   * @private visible for testing
+   * @private
+   * @visibleForTesting
    */
-  invoke_(invocation, actionInfo) {
-    const target = dev().assertElement(invocation.target);
+  invoke_(invocation, opt_actionInfos) {
     const method = invocation.method;
+    const targetType = invocation.targetType;
+
+    // Check that this action is whitelisted (if a whitelist is set).
+    const whitelist = this.queryWhitelist_();
+    if (whitelist) {
+      const id = `${targetType}.${method}`;
```

Comment:
The whitelist feature now supports action targets other than global "AMP", e.g. "amp-form.submit".

Reply:
LGTM for now; I suspect we'll need more work here once we drop the `standalone` requirement for amp-story, since what we really need is to disallow action invocations whose target is a descendant of an `<amp-story>` element (e.g. even though you can normally do `my-video.play`, you can't when that video is part of a story).